metabase/metabase
# Allow defective duplicate fields

This PR implements the changes in steps 1 and 2 from [this comment](https://github.com/metabase/metabase/issues/44659#issuecomment-2192932023):

> 1. Create a migration to add a new boolean column is_defective_duplicate and mark the defective fields.
> 2. Expand the unique constraint on metabase_field to allow inactive defective duplicate fields to co-exist with the actual active fields of the same name, i.e. a unique constraint on (name, parent_id, table_id, is_defective_duplicate). Also update toucan select hooks to dissoc is_defective_duplicate. After this point, load-from-h2 works again.

It unblocks `load-from-h2` from the problem described in https://github.com/metabase/metabase/issues/44659, which will downgrade the issue from P1 to P2.

**Important background:**

- On all db types, we have a unique index on `metabase_field(table_id, parent_id, name)`. But most fields have `parent_id = null` (this column is only used for nested fields on MongoDB). In most databases, `NULL != NULL`, so that’s why you can have multiple a field with the same `(table_id, name)` and not violate the unique index.
- On Postgres, we made tighter constraints around duplicates because it supports partial index and we added a unique index on `(table_id, name) where parent_id is null`. This is why `load-from-h2` only fails when you migrate to Postgres.

Related: 
- [Slack discussion about how this is blocking load-from-h2.](https://metaboat.slack.com/archives/C013VC33N95/p1718877409134089?thread_ts=1718814222.803569&cid=C013VC33N95)
- [Slack thread started by Ngoc on the solution](https://metaboat.slack.com/archives/C0641E4PB9B/p1719380668675549)
- [Original issue that was closed when the root cause was fixed](https://github.com/metabase/metabase/issues/44459)

### To verify:
1. Start Metabase from https://github.com/metabase/metabase/pull/43812 with `git checkout 33df115`. Use MB with a fresh H2 app DB.
3. Create a Postgres DB with this table
```
DROP DATABASE IF EXISTS test_db; 
CREATE DATABASE test_db;

CREATE TABLE my_table (
    id serial PRIMARY KEY,
    data jsonb
);

INSERT INTO my_table (data)
VALUES ('[1, 2, 3]'::jsonb);
```
4. Sync the created DB. 
5. Verify that defective duplicate fields exist by trying to create a question that selects two "Data" fields from `my_table`.
6. Execute the `load-from-h2` command to migrate the app DB from H2 to Postgres:
```
export MB_DB_TYPE=postgres
export MB_DB_CONNECTION_URI="jdbc:postgresql://<host>:5432/metabase?user=<username>&password=<password>"
clojure -M:run load-from-h2 /path/to/metabase.db
```
It should not succeed.
7. Checkout this branch and try step 6 again. It should succeed this time.
8. Start Metabase again using the newly migrated Postgres app DB.
```
export MB_DB_TYPE=postgres
export MB_DB_CONNECTION_URI="jdbc:postgresql://<host>:5432/metabase?user=<username>&password=<password>"
clojure -M:dev
```
